File: Blazor/BlazorApp/Models/ComplexNum.cs
namespace BlazorApp.Models
{
    public class ComplexNum
    {
        public double Real
        {
            get;
            set;
        }

        public double Imaginary
        {
            get;
            set;
        }

        public double AbsoluteValue
        {
            get
            {
                var discriminant = Math.Pow(this.Real, 2) + Math.Pow(this.Imaginary, 2);
                return Math.Sqrt(discriminant);
            }
        }

        public ComplexNum Square
        {
            get
            {
                return new ComplexNum()
                {
                    Real = this.Real * this.Real + (this.Imaginary * this.Imaginary * -1),
                    Imaginary = this.Real * this.Imaginary + this.Real * this.Imaginary
                };
            }
        }
        public static ComplexNum operator +(ComplexNum c1, ComplexNum c2) => new ComplexNum() { Real = c1.Real + c2.Real, Imaginary = c1.Imaginary + c2.Imaginary };
    }
}

File: Ember/emberapp/app/models/complex-num.js
export class Complex {
  constructor(real, imaginary) {
    this.real = real;
    this.imaginary = imaginary;
  }
}

export function addComplex(c1, c2) {
  var real = c1.real + c2.real;
  var imaginary = c1.imaginary + c2.imaginary;
  return new Complex(real, imaginary);
}

export function square(c1) {
  var real = c1.real * c1.real + c1.imaginary * c1.imaginary * -1;
  var imaginary = c1.real * c1.imaginary + c1.imaginary * c1.real;

  return new Complex(real, imaginary);
}

export function abs(c) {
  var discriminant = Math.pow(c.real, 2) + Math.pow(c.imaginary, 2);
  return Math.sqrt(discriminant);
}

